### ISSUE TITLE
Nginx Timeout

### ISSUE THREAD

**Issue by pixeltherapy** [afragen/github-updater#532]
I've run into an interesting problem using GHU on Plesk Onyx. When installing/updating a 'large' (18MB) theme from a private GitHub repo the connection will time-out. Smaller themes and plugins work fine. 

GHU will just sit there until it errors with a 503/504 or Chrome returns an `ERR_INCOMPLETE_CHUNKED_ENCODING`. 
I'm running PHP7.x as PHP-FPM via nginx proxy which will return the error: 
`upstream timed out (110: Connection timed out) while reading upstream`
If I try running as FastCGI the error is: 
`mod_fcgid: read data timeout...`
This testing shows that it is indeed my nginx upstream proxy that is cutting the connection at whatever timeout I set (60-600 sec).

Now Plesk is a bit of a beast to tweak as regards timeouts but I have upped all the relevant nginx and PHP timeouts/keepalives for testing but to no avail. There is the argument that I shouldn't have to increase timeouts or keepalives willy-nilly, which is sage advice, and I doubt you wished people to edit nginx.conf to get the plugin running.

Using a similarly heavy process, which is the plugin BackupBuddy, will reveal the same error when zipping a large file (500MB) but this can be overcome by upping the PHP max_exec time temporarily. The same solution does not work for GHU so the issue is elsewhere.

I'm posting this as an issue as GHU is the only script that has this problem, otherwise my server is blazing along :o) My hope is that there is some way within GHU to improve the keepalive aspect or improve process logging to help identify where the error lies.

I'm here for any more info if you need it, it's just that there's a lot of info on this one ;o)

**Comment by afragen** [comment]
This really doesn't have anything to do with GitHub Updater as much as it has to do with large file transfers and your connection speed. 😉 
GitHub Updater only uses core WordPress functions for updating.

Any solution to this is going to come on the server side.

In `ngnix.conf` you should be able to increase the timeout by modifying or adding the following line.

```
add fastcgi_read_timeout 3600s;
```
https://www.fdgweb.com/increase-php-timeout-values-apache-ngnix/

**Comment by pixeltherapy** [comment]
I figured that was the case but without any difficulty with large tasks outside GHU I thought there may be some 'black magic' in how GHU performs the update outside of core WP functions that might shed light on my problem. At the time I was too fried to read GHU source myself :flushed:

The `nginx.conf` config you suggest was already tried but I'm going at it again later this week so do leave this issue open for the moment and I will post my findings. It may help others hitting the same wall.

PS: that's some timeout suggestion 😅  I like your style!

**Comment by afragen** [comment]
I use that timeout locally so when I step through code I don't error. On a live server 5 minutes is probably more than enough. 🤓

**Comment by pixeltherapy** [comment]
So yes, I went back to debug the issue and I haven't been able to replicate it. Everything is 'mysteriously' fine without messing with any nginx directives or similar faffing about. The same private repo release updates just fine on a number of instances on the same server.

Now I did make some other optimisations to the server but nothing that should have resolved the issue I was having. I'm hardly that good.

So, apart from some unknowable intermittent issue in the pipes between my server and github I can only guess that it was a file permissions problem that manifested itself as an fcgi timeout.

Consider the issue resolved, or non-existent 😬!
